src/client/easy.c

```diff
@@ -0,0 +1,780 @@
+ *
+ * @a\017\0voter\0@b\0890\0standby\0
+ */
+
+/* Called at startup to parse the node store read from disk into an in-memory
+ * representation. Aborts the process on error, since there's nothing much we
+ * can do with a corrupt node store. The user can respond by removing that file
+ * and then trying again. */
+static void parseNodeStore(const char *buf,
+			   size_t buf_len,
+			   struct node_store_cache *cache)
+{
+	const char *p;
+	size_t n;
+	size_t k;
+	size_t i;
+	const char *addr;
+	const char *id_str;
+	unsigned long long id;
+	const char *role_str;
+	int role;
+	struct client_node_info info;
+
+	p = buf;
+	n = buf_len;
+	while (n > 0) {
+		addr = p;
+		k = strnlen(p, n);
+		if (k == n) {
+			fatal("dqlite_server: parseNodeStore");
+		}
+		p += k + 1;
+		n -= k + 1;
+
+		id_str = p;
+		k = strnlen(p, n);
+		if (k == n) {
+			fatal("dqlite_server: parseNodeStore");
+		}
+		p += k + 1;
+		n -= k + 1;
+		/* Be stricter than strtoull: digits only! */
+		for (i = 0; i < k; i += 1) {
+			if (id_str[i] < '0' || id_str[i] > '9') {
+				fatal("dqlite_server: parseNodeStore");
+			}
+		}
+		errno = 0;
+		id = strtoull(id_str, NULL, 10);
+		if (errno != 0) {
+			fatal("dqlite_server: parseNodeStore");
+		}
+
+		role_str = p;
+		k = strnlen(p, n);
+		if (k == n) {
+			fatal("dqlite_server: parseNodeStore");
+		}
+		p += k + 1;
+		n -= k + 1;
+		if (strcmp(role_str, "spare") == 0) {
+			role = DQLITE_SPARE;
+		} else if (strcmp(role_str, "standby") == 0) {
+			role = DQLITE_STANDBY;
+		} else if (strcmp(role_str, "voter") == 0) {
+			role = DQLITE_VOTER;
+		} else {
+			fatal("dqlite_server: parseNodeStore");
+		}
+
+		info.addr = strdupChecked(addr);
+		info.id = (uint64_t)id;
+		info.role = role;
+		pushNodeInfo(cache, info);
+	}
+}
+
+/* Write the in-memory node store to disk. This discards errors, because:
+ *
+ * - we can't do much to handle any of these error cases
+ * - we don't want to stop everything when this encounters an error, since the
+ *   persisted node store is an optimization, so it's not disastrous for it to
+ *   be missing or out of date
+ * - there is already a "retry" mechanism in the form of the refreshTask thread,
+ *   which periodically tries to write the node store file
+ */
+static void writeNodeStore(struct dqlite_server *server)
+{
+	int store_fd;
+	FILE *f;
+	unsigned i;
+	ssize_t k;
+	const char *role_name;
+	int rv;
+
+	store_fd = openat(server->dir_fd, "node-store-stale",
+			  O_RDWR | O_CREAT | O_TRUNC, 0644);
+	if (store_fd < 0) {
+		return;
+	}
+	f = fdopen(store_fd, "w+");
+	if (f == NULL) {
+		close(store_fd);
+		return;
+	}
+
+	for (i = 0; i < server->cache.len; i += 1) {
+		role_name =
+		    (server->cache.nodes[i].role == DQLITE_SPARE)
+			? "spare"
+			: ((server->cache.nodes[i].role == DQLITE_STANDBY)
+			       ? "standby"
+			       : "voter");
+		k = fprintf(f, "%s%c%" PRIu64 "%c%s%c",
+			    server->cache.nodes[i].addr, '\0',
+			    server->cache.nodes[i].id, '\0', role_name, '\0');
```

Comment:
The format of the node store file seems a mix of binary format and text format. It's binary because it's a sequence of 0-terminated strings, but it's also text because some values like the ID and the role are converted to strings.

What about having a fully textual format, that one can easily inspect with `cat`?

Probably something like:

```C
k = fprintf(f, "%s%c%" PRIu64 "%c%s%c",
			    server->cache.nodes[i].addr, ' ',
			    server->cache.nodes[i].id, ' ', role_name, '\n');
```

This would require addresses to have no whitespaces in them, which is probably a reasonable constraint.

There should also be some kind of versioning of the file, something as simple as `format 1` line at the beginning of the file would be enough. Or anything equivalent.

Reply:
FWIW the current binary format of dqlite was designed to minimize encoding and memory copy costs: it takes advantage of memory alignment and native binary representation (e.g. for numbers), a bit like https://capnproto.org/ and other zero-copy serialization formats. Performance is of course not relevant here, but since we already have a format and tooling for it I think it'd make sense to use it.

The internal C code/tooling that we have to support this format can be improved if desired, but at the end of the day all encoders are a variation of `my_encode(struct my_object *obj, void **buf /* out */)` and decoding is the way round.

The NUL-terminated format is surely handy for C, probably a bit less for other languages. If you don't want to use the format we already, I'd still suggest to turn this into a newline-terminated (`\n` instead of `0`) which at least has the advantage to be fully textual and more friendly to scripting languages.